tests/returntypes/test_jax_new.py

```diff
@@ -0,0 +1,1125 @@
+
+            with tape:
+                qml.RX(a, wires=0)
+                U3(*p, wires=0)
+                qml.expval(qml.PauliX(0))
+
+            tape = tape.expand(stop_at=lambda obj: device.supports_operation(obj.name))
+            return execute([tape], device, interface=interface, **execute_kwargs)[0]
+
+        a = jnp.array(0.1)
+        p = jnp.array([0.1, 0.2, 0.3])
+
+        dev = qml.device("default.qubit", wires=1)
+        res = cost_fn(a, p, device=dev)
+        expected = np.cos(a) * np.cos(p[1]) * np.sin(p[0]) + np.sin(a) * (
+            np.cos(p[2]) * np.sin(p[1]) + np.cos(p[0]) * np.cos(p[1]) * np.sin(p[2])
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        jac_fn = jax.grad(cost_fn, argnums=(1))
+        res = jac_fn(a, p, device=dev)
+        expected = jnp.array(
+            [
+                np.cos(p[1]) * (np.cos(a) * np.cos(p[0]) - np.sin(a) * np.sin(p[0]) * np.sin(p[2])),
+                np.cos(p[1]) * np.cos(p[2]) * np.sin(a)
+                - np.sin(p[1])
+                * (np.cos(a) * np.sin(p[0]) + np.cos(p[0]) * np.sin(a) * np.sin(p[2])),
+                np.sin(a)
+                * (np.cos(p[0]) * np.cos(p[1]) * np.cos(p[2]) - np.sin(p[1]) * np.sin(p[2])),
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_independent_expval(self, execute_kwargs, interface):
+        """Tests computing an expectation value that is independent of trainable
+        parameters."""
+        dev = qml.device("default.qubit", wires=2)
+        params = jnp.array([0.1, 0.2, 0.3])
+
+        def cost(a, cache):
+            with qml.tape.QuantumTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.RY(a[2], wires=0)
+                qml.expval(qml.PauliZ(1))
+
+            res = execute([tape], dev, cache=cache, interface=interface, **execute_kwargs)
+            print(res)
+            return res[0]
+
+        res = jax.grad(cost)(params, cache=None)
+        assert res.shape == (3,)
+
+
+@pytest.mark.parametrize("execute_kwargs", execute_kwargs)
+class TestVectorValued:
+    """Test vector-valued jacobian returns for the JAX Python interface."""
+
+    def test_multiple_expvals(self, execute_kwargs):
+        """Tests computing multiple expectation values in a tape."""
+
+        dev = qml.device("default.qubit", wires=2)
+        params = jnp.array([0.1, 0.2, 0.3])
+
+        def cost(a, cache):
+            with qml.tape.QuantumTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.RY(a[2], wires=0)
+                qml.expval(qml.PauliZ(0))
+                qml.expval(qml.PauliZ(1))
+
+            res = qml.interfaces.execute(
+                [tape], dev, cache=cache, interface="jax-python", **execute_kwargs
+            )
+            return res[0]
+
+        res = jax.jacobian(cost)(params, cache=None)
```

Comment:
Could we double-check in a test case that using both `jacrev` and `jacfwd` works well?

Reply:
Very good idea, I will add those.